docs/okeps/okep-5094-layer2-transit-router.md

```diff
@@ -0,0 +1,991 @@
+- Pass transit_router as cluster_name to the gateway init functions
+- Remove from gateway.go the code that was attaching GR to layer2 switch
+- Add to gateway.go the code that connects GR to transit_router using the peer ports
+- Change egressip.go so routes are configured at transit_router instead of gateway router.
+- Change the ZoneInterConnect handler to be able to add the transit router remote ports to transit_router
+- In general re-use as much Layer3 code as possible since these makes Layer2 topology similar to it.
+
+#### At ovnkube-control-plane
+
+At cluster manager at the zoneCluster controller calculate the peers node subnet
+and annotate the node.
+
+To calculate the peer node subnet we have two options:
+- Derive it from the node-id (same we do for transit switch addrs):
+  - good:
+    - No need to maintain the allocator lifecycle
+    - It's possible to entertain the idea of not annotating the node since the source of truth is the node-id
+  - bad:
+    - Add complexity in the form of node-id to subnet derivation code
+    - Make the transit peers node subnet dependent on node-id
+- Create a new subnet allocator that will re-sync on restarts and allocate
+  the subnet with it:
+  - good:
+    - Is simpler since it's matter of a new subnet allocator
+    - The transit peers node subnet is no longer dependant on node-id
+  - bad:
+    - Need to maintain the allocator lifecycle
+    - Use more memory
+
+We will go with the first option to avoid polluting already overloaded node annotations.
+
+### Rolling upgrade and traffic disruption
+
+Openshift upgrades work as follows: first the ovn-k pods are upgraded while the 
+workload pods are still running. We can't upgrade topology at this point, 
+because it includes SNAT for the management port move from the GR to the transit router, which is disruptive 
+for existing connections. Some time after that the node is drained (no more workload pods are left) 
+and rebooted, at this point ovn-k is restarted with no workload pods, which is the time 
+when we can make the topology upgrade.
+
+From the ovn-k side we need to figure out when is that time with no workload pods, so we will 
+introduce a function similar to https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5416/commits/8adda434a35831c49e9bd19b86888bfb074be89f#diff-214cf3919602fd060047b8d15fd6c0ca9d3ed3d42c47fff4b181a072c182b673R306 
+that will check whether a given network has already been updated and if not if it has running pods, 
+and only upgrade the topology when it doesn't.
+This means that we will have to leave the code for the previous topology version in place for one more release, 
+and only cleanup afterwards.
+
+This means, we don't need to worry about existing connections disruption (since the topology upgrade happens
+after the node reboot, so there are no running connections), but we need to make sure that new connections
+created when some nodes are upgraded and some are not will work and won't be disrupted.
+
```

Comment:
Since in this version of ovnk we will support upgrading from non-IC -> IC, are there any concerns about upgrading a layer 2 network while non-IC-> IC is also upgrading? I think UDN was not supported pre-IC, and also layer 2 as primary was also not supported right? But layer 2 secondary NAD was supported. It would be good to clarify this explicitly here.